### PR TITLE
TELCODOCS-649: NUMA GA readiness, removing dev previews and docs update

### PR DESCRIPTION
**Pull request by rohennes**
TELCODOCS#649: Remove TP for GA in 4.13 of NUMA operator. Some updates to content for GA.

Version(s):
4.13+

Issue:
- https://issues.redhat.com/browse/TELCODOCS-649
- https://issues.redhat.com/browse/TELCODOCS-975

Link to docs preview:
- Edited sentence in step 4 to clarify pod resource calculation:
https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-scheduling-numa-aware-workloads_numa-aware

- Edited scheduler YAML and added note for `cacheResyncPeriod`:
https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-deploying-the-numa-aware-scheduler_numa-aware

- Added note for PodResources API:
https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-about-numa-aware-scheduling_numa-aware

- New topic for configuring polling options:
https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-configuring-node-groups-for-the-numaresourcesoperator_numa-aware

 - https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html
 

 
 
 
QE review:
- [x] QE has approved this change.

Additional information:
This PR addresses two TELCODOCS stories as the changes for 975 build on top of the changes for 649.

### CODE REVIEW

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+:_content-type: PROCEDURE
+
+[id="cnf-configuring-node-groups-for-the-numaresourcesoperator_{context}"]
+= Configuring NUMA Resource Operator polling
+
+You can configure how the NUMA Resources Operator polls a `nodeGroup` for available NUMA resources by configuing the `spec.nodeGroups` specification in the `NUMAResourcesOperator` custom resource (CR). The configuration options are the following:
+
+* `infoRefreshMode`: Determines the trigger for polling updates to the API Server.
+* `infoRefreshPeriod`: Determines the duration between polling updates.
+* `podsFingerprinting`: Determines if point-in-time information for the current set of pods running on a node is exposed in polling updates.
```
Comment (rohennes) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:14]: I am a little unclear about what information is gathered in pod finger printing. Perhaps need to some help adding more detail?

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 2]
```diff
@@ -0,0 +1,72 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
```
Comment (rohennes) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:4]: I added this as a separate procedure. However, it's also an option to just edit the existing YAML for [Creating the NUMAResourcesOperator custom resource](https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-creating-nrop-cr_numa-aware)  to include these specifications. What is the benefit of fine-tuning the polling exactly and do you think it justifies it's own topic?

**Comment by ffromani** [comment]
> **Questions:**
> 
>     * Need to make changes to other yaml in these sections regarding the API version? Or a note regarding compatibility between versions?

I'd do change indeed the API version in the example YAMLs for 4.13 and beyond (early warning: we may need to change this for 4.12 in the future)

>      * Need to add info for troubleshooting the scheduling plugin cache?

Nice to have, if we have time and capacity, yes.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+:_content-type: PROCEDURE
+
+[id="cnf-configuring-node-groups-for-the-numaresourcesoperator_{context}"]
+= Configuring NUMA Resource Operator polling
+
+You can configure how the NUMA Resources Operator polls a `nodeGroup` for available NUMA resources by configuing the `spec.nodeGroups` specification in the `NUMAResourcesOperator` custom resource (CR). The configuration options are the following:
```
Comment (ffromani) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:10]: typo: `configuring`

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+* `infoRefreshMode`: Determines the trigger for polling updates to the API Server.
```
Comment (ffromani) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:12]: I'm not a native english speaker, so bear with me: this seems to suggest somehow the updates could be done polling the apiserver (which is something a kubernetes component CAN do in theory!).  I'd rephrase in a different way in order to convey all the following
1. updates are done polling _the kubelet_
2. the trigger condition for the polling is controlled by this option
3. updates are thus _sent_ to the apiserver, regardless to the polling trigger

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+* `infoRefreshPeriod`: Determines the duration between polling updates.
+* `podsFingerprinting`: Determines if point-in-time information for the current set of pods running on a node is exposed in polling updates.
++
+[NOTE]
+====
+You must enable `podsFingerprinting` to facilitate the `cacheResyncPeriod` specification in the `NUMAResourcesScheduler`, which helps to report more exact resource availability by monitoring pending resources on nodes.
```
Comment (ffromani) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:18]: this is correct, but I'll take the opposite direction, explaining that:
`podsFingerprint is required for `cacheResyncPeriod` (the latter won't work without `podsFingerprint`)
`podsFingerprint` is enabled by default, but you can opt-out by...

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc [hunk 6]
```diff
@@ -78,7 +78,14 @@ metadata:
   name: numaresourcesscheduler
 spec:
   imageSpec: "registry.redhat.io/openshift4/noderesourcetopology-scheduler-container-rhel8:v{product-version}"
+  cacheResyncPeriod: "5s" <1> 
 ----
+<1> Optional: Enter an interval value in seconds for synchronization of the scheduler cache.
```
Comment (ffromani) [inline modules/cnf-deploying-the-numa-aware-scheduler.adoc:83]: this is technically optional because implementation details but it's strongly strongly recommended to be enabled, I'd emphasize this in the documentation. Let's suggest the `5s` default, further tuning information will be deferred to knowledge base articles.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc (hunk 2 above)
Comment (ffromani) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:4]: Fine-tuning is meant to be exposed as troubleshooting and/or advanced settings. The expectations is the default would be fine, and only specific conditions will require fine-tuning.

File: modules/cnf-about-numa-aware-scheduling.adoc [hunk 7]
```diff
@@ -21,3 +21,8 @@ NodeResourceTopology API:: The `NodeResourceTopology` API describes the availabl
 NUMA-aware scheduler:: The NUMA-aware secondary scheduler receives information about the available NUMA zones from the `NodeResourceTopology` API and schedules high-performance workloads on a node where it can be optimally processed.
 Node topology exporter:: The node topology exporter exposes the available NUMA zone resources for each compute node to the `NodeResourceTopology` API. The node topology exporter daemon tracks the resource allocation from the kubelet by using the `PodResources` API.
 PodResources API:: The `PodResources` API is local to each node and exposes the resource topology and available resources to the kubelet.
++
+[NOTE]
+====
+The `List` endpoint of the `PodResources` API exposes exclusive CPUs allocated to a particular container. The API does not expose CPUs that belong to a shared pool.
```
Comment (swatisehgal) [inline modules/cnf-about-numa-aware-scheduling.adoc:27]: Since we are highlighting the `List` endpoint, perhaps we should mention the `GetAllocatableResources` endpoint as well given that we use that to learn about allocatable resource. You can refer to the related [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments) here for details.

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc (hunk 6 above)
Comment (rohennes) [reply modules/cnf-deploying-the-numa-aware-scheduler.adoc:83]: I've removed the word optional at the start. I guess the note says it is disabled by default so that's enough indication that this is optional.
We try to avoid "recommending" anything in the docs. If we recommend an option, it can leave Red Hat open to blame if something goes wrong later. So I've tried to re-word this sentence another way. Hope it works for you.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc (hunk 2 above)
Comment (rohennes) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:4]: Thanks for the info @ffromani. I think it is correct then that this is a separate procedure from the default flow. However, I'd like to include the "why" in the opening paragraph. Why would a user do this? What's the benefit? 

Something like:
"Fine-tuning the polling operations of the NROP allows you to minimize network traffic while maintaining the highest accuracy information of available NUMA resources." Or whatever the ultimate design goal is...  
WDYT?

Comment (ffromani) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:4]: We anticpate fine tuning to be useful in two scenarios:
1. depending on the unpredictable combination of cluster layout (how many nodes...) and workload, improve the scheduling behavior in terms of scheduled pod throughput vs scheduler effectiveness (remove/minimize unexpected scheduling failures)
RATIONALE: we strive to provide good defaults, but "one size does not fit all" is especially true for this feature
2. debug and troubleshoot bad scheduling decisions, which is a variation of the above
3. it's indeed possible to further optimize the load between components and control plane with fine tuning, but we don't anticipate this to happen frequently yet.

In a nutshell, I'd expect fine tuning to be needed to troubleshoot failures and/or when defaults are ineffective.

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc (hunk 6 above)
Comment (ffromani) [reply modules/cnf-deploying-the-numa-aware-scheduler.adoc]: Thanks for the explanation. I'll read again but the gist is that: this is disabled by default because historical reasons which we couldn't help, but every user should opt-in unless they have strong reasons not to. We can't just enable it by default, otherwise we would have done it.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 8]
```diff
@@ -0,0 +1,72 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_content-type: PROCEDURE
+
+[id="cnf-configuring-node-groups-for-the-numaresourcesoperator_{context}"]
+= Configuring NUMA Resource Operator polling
```
Comment (ffromani) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:8]: the subject here is wrong: the numaresources operator in turn configures agents on the node (daemonsets, resource-topology-exporter) which in turn do the polling according to the settings documented here.
Everything else looks fine.

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc [hunk 9]
```diff
@@ -72,13 +72,22 @@ $ oc create -f nro-kubeletconfig.yaml
...
   imageSpec: "registry.redhat.io/openshift4/noderesourcetopology-scheduler-container-rhel8:v{product-version}"
+  cacheResyncPeriod: "5s" <1> 
 ----
+<1> Enter an interval value in seconds for synchronization of the scheduler cache. A value of `5s` is typical for most implementations.
++
+[NOTE]
+====
+The `cacheResyncPeriod` specification is disabled by default. Enable to help the NUMA Resource Operator report more exact resource availability by monitoring pending resources on nodes and synchronizing this information in the scheduler cache at a defined interval. This also helps to minimize `Topology Affinity Error` errors because of sub-optimal scheduling decisions. The lower the interval the greater the network load.
+
+The `podsFingerprinting` specification in the `NUMAResourcesOperator` CR is a requirement for the implementation of the `cacheResyncPeriod` specification.
```
Comment (ffromani) [inline modules/cnf-deploying-the-numa-aware-scheduler.adoc:89]: the podsFingerprinting=true specification... ?

**Review by ffromani — COMMENTED** [review]
Almost there, we probably need to clarify further a couple points and then it should be good.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc (hunk 2 above)
Comment (rohennes) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:4]: Appreciate the info @ffromani. I've added some of this to the opening paragraph. I think it gives better context to the user about when they should use this feature.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 10]
```diff
@@ -0,0 +1,72 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_content-type: PROCEDURE
+
+[id="cnf-configuring-node-groups-for-the-numaresourcesoperator_{context}"]
+= Configuring polling operations for NUMA resource updates
+
+You can configure how the NUMA Resources Operator polls a `nodeGroup` for available NUMA resources by configuring the `spec.nodeGroups` specification in the `NUMAResourcesOperator` custom resource (CR). This provides advanced control of NUMA Resources Operator polling operations. Configure these specifications to improve scheduling behaviour and troubleshoot suboptimal scheduling decisions. The configuration options are the following:
```
Comment (ffromani) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:10]: this looks good _except_ the fact that:
1. the NUMA Resources Operator does watch the `spec.nodeGroups` in the `NUMAResorucesOperator` CR but
2. the NUMA Resources Operator USES this data to set the configuration of the daemonset pertaining to that nodeGroup
3. the daemons governed by the daemonset will in turn poll the kubelets on their nodes

So "NUMAResources polls a nodeGroup" is wrong.

**Review by ffromani — COMMENTED** [review]
I think we need to clarify an item and this PR is ready.

**Comment by ffromani** [comment]
/lgtm

thanks for the updates

**Review by swatisehgal — APPROVED** [review]
/lgtm

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc [hunk 11]
```diff
@@ -8,62 +8,50 @@
...
 +
 [source,yaml]
 ----
-apiVersion: machineconfiguration.openshift.io/v1
-kind: KubeletConfig
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
 metadata:
-  name: cnf-worker-tuning
+  name: perfprof-nrop
```
Comment (rohennes) [inline modules/cnf-deploying-the-numa-aware-scheduler.adoc:35]: @ffromani - I just placed the `annotations` content as a placeholder now. Can you provide some details about the known issue and what the fix does? 
I think it's OK to just include fix as commented lines in this YAML and the user can un-comment if they are encountering the issue?

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc [hunk 12]
```diff
@@ -8,62 +8,50 @@
 
 After you install the NUMA Resources Operator, do the following to deploy the NUMA-aware secondary pod scheduler:
```
Comment (rohennes) [inline modules/cnf-deploying-the-numa-aware-scheduler.adoc:10]: @ffromani - I removed the bullet here saying you need to create a MCP and a pod admittance policy. 

I think the pod admittance policy is not needed. And the MCP was created during the procedure to create the NUMAResourcesOperator custom resource. So instead, I moved the task to create an MCP as a prerequisite.

Comment (ffromani) [reply modules/cnf-deploying-the-numa-aware-scheduler.adoc:10]: "pod admittance policy" is one of the things we can configure using a performance profile. I think the latest version you uploaded is better.

**Review by ffromani — COMMENTED** [review]
/lgtm

iff the `MachineConfigPool` creation step is only related to the manual configuration, the flow without `PerformanceProfile`. Even in that case we may want to document that the specific `worker` pool exists by default.

File: modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+.Prerequisites
+
+* Install the OpenShift CLI (`oc`).
+* Log in as a user with `cluster-admin` privileges.
+* Install the NUMA Resources Operator.
+
+.Procedure
+
+. Create the `MachineConfigPool` custom resource that enables custom kubelet configurations for worker nodes:
```
[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc:19]: I suggest the following: if the `MachineConfigPool` step was present in the existing doc, let's leave it as is in the manual flow. If was added for the sake of completeness, I'd re-name the MachineConfigPool like `worker-numa` or something (not use the `worker` name) and emphasize that is an example, thus optional.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+$ oc get numaresop numaresourcesoperator -o json | jq '.status'
+----
++
+.Example output
+[source,terminal]
+----
+      ...
+
+        "config": {
+        "infoRefreshMode": "Periodic",
```
Comment (kononovn) [inline modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:66]: I've managed to exec cmd above:
`oc get numaresop numaresourcesoperator -o json | jq '.status'`
My output is a bit different:
      "config": {
        "infoRefreshMode": **"PeriodicAndEvents",**
        "infoRefreshPeriod": "10s",
        "podsFingerprinting": "Enabled"
      },
      "name": "worker"

Comment (ffromani) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:66]: numaresources operator defaults: PeriodicAndEvents
recommended configuration: Periodic

We will evaluate if we can change the defaults, but is not easy.

Comment (ffromani) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:66]: > We will evaluate if we can change the defaults, but is not easy.

We can't. So the manual configuration part will need to include this bit. The user needs anyway to specify the `NodeGroup` (or in general is required to send configuration to the operator). so it should not be a big deal

File: modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc:19]: sounds good to me. Keep it in the manual flow (old flow now demoted) and don't add in the new and recommended flow involving performanceprofile.

File: modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc (hunk 14 above)
Comment (rohennes) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:66]: I think the configuration of the NUMAResourcesOperator CR is detailed in this section. And the example output is a reflection of that configuration. So I don't think any changes are required? or I could add a line to say the default is PeriodicAndEvents?

Comment (ffromani) [reply modules/cnf-configuring-node-groups-for-the-numaresourcesoperator.adoc:66]: @rohennes I agree with you. If the example output reflects the expected outcome of the provided CR, no changes are needed.

**Comment by ffromani** [comment]
/lgtm

I think we're good now

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+[NOTE]
+====
+Deployments that request more resources than is available for scheduling will fail with a `MinimumReplicasUnavailable` error. The deployment succeeds when the required resources become available. Pods remain in the `Pending` state until the required resources are available.
+====
+
+. Verify that the expected allocated resources are listed for the node. Run the following command:
++
+[source,terminal]
+----
+$ oc describe noderesourcetopologies.topology.node.k8s.io
```
Comment (kononovn) [inline modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:136]: In deployment manifest you specify replica:1 which means only 1 deployment pod. 
Before running this cmd you need to check on what Node this pod is running:
oc get pods -n debug -o wide
NAME                               READY   STATUS    RESTARTS   AGE     IP            NODE       NOMINATED NODE   READINESS GATES
numa-deployment-1-958ddbfc-grcqc   2/2     Running   0          6m55s   10.128.2.50   **worker-0**   <none>           <none>


 on then run following cmd:

oc describe noderesourcetopologies.topology.node.k8s.io **worker-0**

**Review by kononovn — CHANGES_REQUESTED** [review]
Overall looks good.Please check few comments that I've added.

File: modules/cnf-deploying-the-numa-aware-scheduler-with-manual-performance-settings.adoc [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  imageSpec: "registry.redhat.io/openshift4/noderesourcetopology-scheduler-container-rhel8:v{product-version}"
+  cacheResyncPeriod: "5s" <1> 
+----
+<1> Enter an interval value in seconds for synchronization of the scheduler cache. A value of `5s` is typical for most implementations.
++
+[NOTE]
+====
+* Enable the `cacheResyncPeriod` specification to help the NUMA Resource Operator report more exact resource availability by monitoring pending resources on nodes and synchronizing this information in the scheduler cache at a defined interval. This also helps to minimize `Topology Affinity Error` errors because of sub-optimal scheduling decisions. The lower the interval the greater the network load. The `cacheResyncPeriod` specification is disabled by default.
+
+* Setting a value of `true` for the `podsFingerprinting` specification in the `NUMAResourcesOperator` CR is a requirement for the implementation of the `cacheResyncPeriod` specification.
```
[the remark this answers is not in the document]
Comment (kononovn) [reply modules/cnf-deploying-the-numa-aware-scheduler-with-manual-performance-settings.adoc:89]: On my cluster it looks like Enabled:
> oc get NUMAResourcesOperator numaresourcesoperator -o yaml
apiVersion: nodetopology.openshift.io/v1
kind: NUMAResourcesOperator
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nodetopology.openshift.io/v1alpha1","kind":"NUMAResourcesOperator","metadata":{"annotations":{},"name":"numaresourcesoperator"},"spec":{"logLevel":"Trace","nodeGroups":[{"machineConfigPoolSelector":{"matchLabels":{"pools.operator.machineconfiguration.openshift.io/worker":""}}}]}}
  creationTimestamp: "2023-03-14T15:17:53Z"
  generation: 2
  name: numaresourcesoperator
  resourceVersion: "119922"
  uid: f2ed7d28-e778-48aa-aee9-6ce4e15bf73c
spec:
  logLevel: Trace
  nodeGroups:
  - config:
      podsFingerprinting: Enabled
  - machineConfigPoolSelector:
      matchLabels:
        pools.operator.machineconfiguration.openshift.io/worker: ""

[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-deploying-the-numa-aware-scheduler-with-manual-performance-settings.adoc:89]: Good catch! Will change to Enabled

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+          limits:
+            memory: "100Mi"
+            cpu: "10"
+          requests:
+            memory: "100Mi"
+            cpu: "10"
+      - name: ctnr2
+        image: gcr.io/google_containers/pause-amd64:3.0
+        imagePullPolicy: IfNotPresent
+        command: ["/bin/sh", "-c"]
```
[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:72]: I ran just this deployment CR **only** on a 4.13 cluster, seemed to work ok?
`$ oc create -f nro-deployment.yaml 
Warning: would violate PodSecurity "restricted:latest": seccompProfile (pod or containers "ctnr", "ctnr2" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
deployment.apps/numa-deployment-1 created`

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc (hunk 15 above)
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc]: Makes sense, thanks. Updated the PR.

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+.. Save the following YAML in the `nro-deployment.yaml` file:
++
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: numa-deployment-1
+  namespace: openshift-numaresources
```
[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:46]: better use another namespace, even `default`. Or just leave the namespace unspecified in the example (perhaps the best approach)

File: modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+  name: numaresourcesoperator
+spec:
+  nodeGroups:
+  - machineConfigPoolSelector:
+      matchLabels:
+        pools.operator.machineconfiguration.openshift.io/worker: "" <1>
+----
+<1> Should match the label applied to worker nodes in the related `MachineConfigPool` CR.
+
+.. Create the `NUMAResourcesOperator` CR by running the following command:
```
[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc:72]: After discussing with Francesco, I think we will not include these "node reboot" details as it is an implementation detail of MCO and this detail may change over time. (if it doesn't need a reboot in future it would be messy trying to clean up the docs!)

[the remark this answers is not in the document]
Comment (kononovn) [reply modules/cnf-creating-nrop-cr-with-manual-performance-settings.adoc:72]: Ok +1

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc (hunk 17 above)
[the remark this answers is not in the document]
Comment (kononovn) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:72]: I also manage to > oc create -f nro-deployment.yaml
however if you check pod status 
oc get pods -n SOME_NAMESPACE 

You will see that actual pods failed to not start due to invalid command. If you exec: 

oc describe pod POD_NAME -n SOME_NAMESPACE 

You can see this error:
Warning Failed 6m33s kubelet Error: container create failed: time="2023-03-14T19:36:07Z" level=error msg="runc create failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory

[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:72]: I removed the `cmd` and `args` lines as requested, thanks!

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:46]: Updated to use replaceable namespace value

File: modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+topo-aware-scheduler
+----
+
+. Create a `Deployment` CR that uses scheduler named `topo-aware-scheduler`, for example:
+
+.. Save the following YAML in the `nro-deployment.yaml` file:
++
+[source,yaml]
+----
+apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (kononovn) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: You can add:
securityContext: <securityContext> <2>

2. securityContext  refer to security context documentation

[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: this seems a bit overkill, let me review the example

[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: > Starting from 4.12 cluster will never schedule workload without proper SecurityContext. Unfortunatelly the example from our documentation'll not work. I don't insist we need to specify SecurityContext but at least we need to mention this fact. WDYT ?

Wan't the enforcing of PSA postponed to 4.14?

[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: @ffromani - any resolution on the above?

[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: not yet. Need to focus on other matters.

[the remark this answers is not in the document]
Comment (rohennes) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: Hi @ffromani - checking in on this again. 
FYI - docs freeze is in just over two weeks (April 12) so it would be good to start closing this one out soon.

[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: I'll have closure this week. From the current testing I've done so far the security context is nice to add but not yet mandatory, so I'm inclined to defer any update in this area to a knowledge base article.

[the remark this answers is not in the document]
Comment (ffromani) [reply modules/cnf-scheduling-numa-aware-workloads-with-manual-performance-setttings.adoc:42]: > I'll have closure this week. From the current testing I've done so far the security context is nice to add but not yet mandatory, so I'm inclined to defer any update in this area to a knowledge base article.

From what I can tell checking recent snapshots of both the NUMA-aware scheduler and builds of 4.13, I confirm this take. IMO it is OK to add this useful information in a knowledge base article and doesn't need to be in the reference docs just yet.

**Comment by kononovn** [comment]
lgtm

**Comment by ffromani** [comment]
LGTM

**Comment by nalhadef** [comment]
Hi. Just want to confirm this callout numbering is as you want it.

image

https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-deploying-the-numa-aware-scheduler_numa-aware

**Comment by ffromani** [comment]
> Hi. Just want to confirm this callout numbering is as you want it. (This occurs in other documents previewed in this PR.)
> 
> image
> 
> https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-deploying-the-numa-aware-scheduler_numa-aware

From engineering perspective the two fields annotated with the callout are related to each other (and needs to be set in a such a way which is compatible with each other), so it makes sense to me to use the same callout number.

**Comment by jeana-redhat** [comment]
> > Hi. Just want to confirm this callout numbering is as you want it. (This occurs in other documents previewed in this PR.)
> > image
> > https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-deploying-the-numa-aware-scheduler_numa-aware
> 
> From engineering perspective the two fields annotated with the callout are related to each other (and needs to be set in a such a way which is compatible with each other), so it makes sense to me to use the same callout number.

I agree but unfortunately the Customer Portal cannot handle this. We can only use a callout once in a file. Since this applies to both items under `cpu`, in this case I would put the `<1>` on the `cpu` line and make the description clear that "In this section, <blah blah blah>".

@rohennes please adjust this one and also make sure there aren't others that do this in the PR (I did not go through it to check for more)

**Comment by rohennes** [comment]
> > > Hi. Just want to confirm this callout numbering is as you want it. (This occurs in other documents previewed in this PR.)
> > > image
> > > https://55855--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling.html#cnf-deploying-the-numa-aware-scheduler_numa-aware
> > 
> > 
> > From engineering perspective the two fields annotated with the callout are related to each other (and needs to be set in a such a way which is compatible with each other), so it makes sense to me to use the same callout number.
> 
> I agree but unfortunately the Customer Portal cannot handle this. We can only use a callout once in a file. Since this applies to both items under `cpu`, in this case I would put the `<1>` on the `cpu` line and make the description clear that "In this section, [blah blah blah]".
> 
> @rohennes please adjust this one and also make sure there aren't others that do this in the PR (I did not go through it to check for more)
> 
> Edit: said "file" above but meant "code block" :) You probably figured that out tho

Done @jeana-redhat - this was the only instance in the PR.

File: scalability_and_performance/cnf-numa-aware-scheduling.adoc [hunk 21]
```diff
@@ -28,11 +27,27 @@ include::modules/cnf-installing-numa-resources-operator-cli.adoc[leveloffset=+2]
...
 
-include::modules/cnf-deploying-the-numa-aware-scheduler.adoc[leveloffset=+1]
+include::modules/cnf-creating-nrop-cr.adoc[leveloffset=+2]
 
-include::modules/cnf-scheduling-numa-aware-workloads.adoc[leveloffset=+1]
+include::modules/cnf-deploying-the-numa-aware-scheduler.adoc[leveloffset=+2]
+
+.Additional resources
+
+* For more information about creating a performance profile, see xref:../scalability_and_performance/cnf-create-performance-profiles.adoc#cnf-about-the-profile-creator-tool_cnf-create-performance-profiles[About the Performance Profile Creator].
```
Comment (jeana-redhat) [inline scalability_and_performance/cnf-numa-aware-scheduling.adoc:38]:
```suggestion
* xref:../scalability_and_performance/cnf-create-performance-profiles.adoc#cnf-about-the-profile-creator-tool_cnf-create-performance-profiles[About the Performance Profile Creator]
```

**Review by jeana-redhat — COMMENTED** [review]
Really excellent work - especially love that you have included plenty of output examples and verification steps. Noticed one very small formatting thing to fix (additional resources should just be the link and its title with no surrounding text), otherwise LGTM
